x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/NodeTermEnumRequest.java

```diff
@@ -0,0 +1,160 @@
+    private int size;
+    private long timeout;
+    private final QueryBuilder indexFilter;
+    private Set<ShardId> shardIds;
+    private String nodeId;
+    
+
+    public NodeTermEnumRequest(StreamInput in) throws IOException {
+        super(in);
+        // Set the clock running as soon as we appear on a node.
+        nodeStartedTimeMillis = System.currentTimeMillis();
+
+        field = in.readString();
+        string = in.readString();
+        caseInsensitive = in.readBoolean();
+        size = in.readVInt();
+        timeout = in.readVLong();
+        taskStartedTimeMillis = in.readVLong();
+        indexFilter = in.readOptionalNamedWriteable(QueryBuilder.class);
+        nodeId = in.readString();
+        int numShards = in.readVInt();
+        shardIds = new HashSet<>(numShards);
+        for (int i = 0; i < numShards; i++) {
+            shardIds.add(new ShardId(in));
+        }
+    }
+
+    public NodeTermEnumRequest(final String nodeId, final Set<ShardId> shardIds, TermEnumRequest request) {
+        this.field = request.field();
+        this.string = request.string();
+        this.caseInsensitive = request.caseInsensitive();
+        this.size = request.size();
+        this.timeout = request.timeout().getMillis();
+        this.taskStartedTimeMillis = request.taskStartTimeMillis;
+        this.indexFilter = request.indexFilter();
+        this.nodeId = nodeId;
+        this.shardIds = shardIds;        
+        
+        // TODO serialize shard ids
+    }
+
+    public String field() {
+        return field;
+    }
+
+    public String string() {
+        return string;
+    }
+
+    public long taskStartedTimeMillis() {
+        return this.taskStartedTimeMillis;
+    }
+    
+    /** 
+     * The time this request was materialized on a shard
+     * (defaults to "now" if serialization was not used e.g. a local request).
+     */
+    public long shardStartedTimeMillis() {
+        if (nodeStartedTimeMillis == 0) {
+            nodeStartedTimeMillis = System.currentTimeMillis();
+        }
+        return this.nodeStartedTimeMillis;
+    }    
+    
+    public Set<ShardId> shardIds() {
+        return Collections.unmodifiableSet(shardIds);
+    }
+
+    public boolean caseInsensitive() {
+        return caseInsensitive;
+    }
+
+    public int size() {
+        return size;
+    }
+
+    public long timeout() {
+        return timeout;
+    }
+    public String nodeId() {
+        return nodeId;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        super.writeTo(out);
+        out.writeString(field);
+        out.writeString(string);
+        out.writeBoolean(caseInsensitive);
+        out.writeVInt(size);
+        // Adjust the amount of permitted time the shard has remaining to gather terms. 
+        long timeSpentSoFarInCoordinatingNode = System.currentTimeMillis() - taskStartedTimeMillis;
+        assert timeSpentSoFarInCoordinatingNode >= 0;
```

Comment:
I don't think that this assert is legal. The delta could be negative if we spent a long time on the above line for instance.